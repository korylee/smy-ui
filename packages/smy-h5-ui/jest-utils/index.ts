import Vue from 'vue'
import { Wrapper } from '@vue/test-utils'
import { IN_BROWSER } from '@smy-h5/shared'

Vue.config.productionTip = false

export const delay = (time = 0) => new Promise((resolve) => setTimeout(resolve, time))

export function getTouch(el: Element | Window | Document, x: number, y: number) {
  return {
    identifier: Date.now(),
    target: el,
    pageX: x,
    pageY: y,
    clientX: x,
    clientY: y,
    radiusX: 2.5,
    radiusY: 2.5,
    rotationAngle: 10,
    force: 0.5,
  }
}

export function trigger(
  wrapper: Wrapper<Vue> | HTMLElement | Window | Document,
  eventName: string,
  x = 0,
  y = 0,
  offsetX = 0,
  offsetY = 0,
) {
  const el = 'element' in wrapper ? wrapper.element : wrapper
  const touchList = [getTouch(el, x, y)]

  const event = document.createEvent('CustomEvent')
  event.initCustomEvent(eventName, true, true, {})

  Object.assign(event, {
    clientX: x,
    clientY: y,
    offsetX,
    offsetY,
    touches: touchList,
    targetTouches: touchList,
    changedTouches: touchList,
  })
  el.dispatchEvent(event)

  return Vue.nextTick()
}

export async function triggerDrag(el: HTMLElement, relativeX: number, relativeY: number) {
  let x = relativeX
  let y = relativeY
  let startX = 0
  let startY = 0
  if (relativeX < 0) {
    startX = Math.abs(relativeX)
    x = 0
  }
  if (relativeY < 0) {
    startY = Math.abs(relativeY)
    y = 0
  }
  await trigger(el, 'touchstart', startX, startY)
  await trigger(el, 'touchmove', x / 4, y / 4)
  await trigger(el, 'touchmove', x / 3, y / 3)
  await trigger(el, 'touchmove', x / 2, y / 2)
  await trigger(el, 'touchmove', x, y)
  await delay(300)
  await trigger(el, 'touchend', x, y)
}

export function mockConsole(method: keyof Console, fn: any = () => {}) {
  const originMethod = console[method] as any
  console[method] = fn
  return () => {
    console[method] = originMethod
  }
}

export function mockGetBoundingClientRect(rect: Partial<DOMRect>): () => void {
  if (IN_BROWSER) {
    const spy = jest.spyOn(Element.prototype, 'getBoundingClientRect')
    spy.mockReturnValue(rect as DOMRect)
    return () => spy.mockRestore()
  }
  return () => {}
}
