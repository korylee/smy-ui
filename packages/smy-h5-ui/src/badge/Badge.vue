<template>
  <div class="smy-badge">
    <div v-show="!hidden && !dot && hasSlots('icon')" :style="style" class="smy-badge__icon">
      <slot name="icon"></slot>
    </div>
    <slot></slot>
    <div
      v-show="!hidden && (internalContent || dot)"
      v-text="internalContent"
      :style="style"
      :class="{
        'smy-badge__content--dot': dot,
        'smy-badge__content--bubble': !dot && bubble,
      }"
      class="smy-badge__content smy-badge__content--sup"
    ></div>
  </div>
</template>
<script>
import { isNumString, isNumber } from '../_utils/is'
import { convertToUnit } from '../_utils/dom'
import { SlotsMixin } from '../_utils/vue/slots'
import { props } from './props'

const isNum = (num) => isNumber(num) || isNumString(num)

export default {
  name: 'SmyBadge',
  mixins: [SlotsMixin],
  props,
  computed: {
    style({ zIndex, color }) {
      return {
        top: convertToUnit(this.top),
        right: convertToUnit(this.right),
        zIndex,
        background: color,
      }
    },
    internalContent() {
      if (this.dot) return
      const { content, max } = this
      if (isNum(content) && isNum(max)) {
        return max < +content ? `${max}+` : content
      }
      return content
    },
  },
}
</script>

<style lang="less">
@import './badge.less';
</style>
