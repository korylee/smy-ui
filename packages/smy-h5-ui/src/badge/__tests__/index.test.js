import { mount } from '@vue/test-utils'
import example from '../example'
import Badge from '..'
import Vue from 'vue'
import { mockConsole } from '../../../jest-utils'

test('test badge example', () => {
  const reset = mockConsole()
  const wrapper = mount(example)

  expect(wrapper.html()).toMatchSnapshot()
  wrapper.destroy()
  reset()
})

test('test badge plugin', () => {
  Vue.use(Badge)
  expect(Vue.component(Badge.name)).toBeTruthy()
})

test('test badge value and max', async () => {
  const wrapper = mount(Badge, {
    propsData: { value: 72 },
  })

  expect(wrapper.find('.smy-badge').text()).toBe('72')
  await wrapper.setProps({ max: 60 })
  expect(wrapper.find('.smy-badge').text()).toBe('60+')
})

test('test badge hidden', () => {
  const wrapper = mount(Badge, {
    propsData: { hidden: true },
  })
  expect(wrapper.find('.smy-badge__content').isVisible()).toBe(false)
})

test('test badge dot', async () => {
  const wrapper = mount(Badge, {
    propsData: { dot: true, value: 72 },
  })
  expect(wrapper.find('.smy-badge__content--dot').exists()).toBe(true)

  expect(wrapper.find('.smy-badge').text()).toBe('')
})

test('test badge position', async () => {
  const template = `
    <smy-badge>
      <div style="width: 100px; height: 100px;"></div>
    </smy-badge>
  `
  const wrapper = mount({
    components: { SmyBadge: Badge },
    template,
  })
  expect(wrapper.html()).toMatchSnapshot()

  await wrapper.setProps({ position: 'right-bottom' })
  expect(wrapper.html()).toMatchSnapshot()

  await wrapper.setProps({ position: 'left-top' })
  expect(wrapper.html()).toMatchSnapshot()

  await wrapper.setProps({ position: 'left-bottom' })
  expect(wrapper.html()).toMatchSnapshot()
})

test('test badge style', () => {
  const wrapper = mount(Badge, {
    propsData: {
      color: '#6200ea',
      top: '100',
      right: '1000',
      zIndex: 500,
    },
  })

  expect(wrapper.html()).toMatchSnapshot()
})
