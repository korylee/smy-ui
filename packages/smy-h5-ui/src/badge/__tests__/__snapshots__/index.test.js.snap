// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test badge example 1`] = `
"<div class=\\"badge-example\\">
  <div class=\\"app-demo-title\\">基础用法</div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--fixed smy-badge__content--right-top\\">8</span></transition-stub> <span class=\\"title\\">标题</span>
  </div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--fixed smy-badge__content--right-top\\">76</span></transition-stub> <span class=\\"title\\">标题</span>
  </div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--fixed smy-badge__content--right-top\\">NEW</span></transition-stub> <span class=\\"title\\">标题</span>
  </div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--bubble smy-badge__content--fixed smy-badge__content--right-top\\">NEW</span></transition-stub> <span class=\\"title\\">标题</span>
  </div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--dot smy-badge__content--fixed smy-badge__content--right-top\\"></span></transition-stub> <span class=\\"title\\">标题</span>
  </div>
  <div class=\\"app-demo-title\\">最大值</div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--fixed smy-badge__content--right-top\\">9+</span></transition-stub> <span class=\\"title\\">标题</span>
  </div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--fixed smy-badge__content--right-top\\">20+</span></transition-stub> <span class=\\"title\\">标题</span>
  </div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--fixed smy-badge__content--right-top\\">99+</span></transition-stub> <span class=\\"title\\">标题</span>
  </div>
  <div class=\\"app-demo-title\\">不同定位</div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--fixed smy-badge__content--right-top\\">200</span></transition-stub> <span class=\\"title\\">右上</span>
  </div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--fixed smy-badge__content--right-bottom\\">200</span></transition-stub> <span class=\\"title\\">右下</span>
  </div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--fixed smy-badge__content--left-top\\">200</span></transition-stub> <span class=\\"title\\">左上</span>
  </div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--fixed smy-badge__content--left-bottom\\">200</span></transition-stub> <span class=\\"title\\">左下</span>
  </div>
  <div class=\\"app-demo-title\\">是否显示(点击“标题”改变状态)</div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--fixed smy-badge__content--right-top\\">8</span></transition-stub> <span class=\\"title\\">标题</span>
  </div>
  <div class=\\"app-demo-title\\">自定义颜色</div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--fixed smy-badge__content--right-top\\">8</span></transition-stub> <span class=\\"title\\">标题</span>
  </div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--fixed smy-badge__content--right-top\\">76</span></transition-stub> <span class=\\"title\\">标题</span>
  </div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--fixed smy-badge__content--right-top\\">NEW</span></transition-stub> <span class=\\"title\\">标题</span>
  </div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--bubble smy-badge__content--fixed smy-badge__content--right-top\\">NEW</span></transition-stub> <span class=\\"title\\">标题</span>
  </div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--dot smy-badge__content--fixed smy-badge__content--right-top\\"></span></transition-stub> <span class=\\"title\\">标题</span>
  </div>
  <div class=\\"app-demo-title\\">自定义位置</div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--fixed smy-badge__content--right-top\\">200</span></transition-stub> <span class=\\"title\\">标题</span>
  </div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--fixed smy-badge__content--right-top\\" style=\\"top: 40px; right: 20px;\\">200</span></transition-stub> <span class=\\"title\\">标题</span>
  </div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--fixed smy-badge__content--right-top\\" style=\\"top: 20px; right: 40px;\\">200</span></transition-stub> <span class=\\"title\\">标题</span>
  </div>
  <div class=\\"app-demo-title\\">自定义徽标内容</div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--icon smy-badge__content--fixed smy-badge__content--right-top\\"><span class=\\"smy-icon\\" style=\\"transition: transform 0ms;\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" version=\\"1.1\\" width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\"><path d=\\"M12.89,3L14.85,3.4L11.11,21L9.15,20.6L12.89,3M19.59,12L16,8.41V5.58L22.42,12L16,18.41V15.58L19.59,12M1.58,12L8,5.58V8.41L4.41,12L8,15.58V18.41L1.58,12Z\\"></path></svg></span></span></transition-stub> <span class=\\"title\\">标题</span>
  </div>
  <div class=\\"app-demo-title\\">独立存在</div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content\\">8</span></transition-stub>
  </div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content\\">76</span></transition-stub>
  </div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content\\">NEW</span></transition-stub>
  </div>
  <div class=\\"smy-badge smy--box\\">
    <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--bubble\\">NEW</span></transition-stub>
  </div>
</div>"
`;

exports[`test badge position 1`] = `
"<div class=\\"smy-badge smy--box\\">
  <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--fixed smy-badge__content--right-top\\"></span></transition-stub>
  <div style=\\"width: 100px; height: 100px;\\"></div>
</div>"
`;

exports[`test badge position 2`] = `
"<div class=\\"smy-badge smy--box\\" position=\\"right-bottom\\">
  <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--fixed smy-badge__content--right-top\\"></span></transition-stub>
  <div style=\\"width: 100px; height: 100px;\\"></div>
</div>"
`;

exports[`test badge position 3`] = `
"<div class=\\"smy-badge smy--box\\" position=\\"left-top\\">
  <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--fixed smy-badge__content--right-top\\"></span></transition-stub>
  <div style=\\"width: 100px; height: 100px;\\"></div>
</div>"
`;

exports[`test badge position 4`] = `
"<div class=\\"smy-badge smy--box\\" position=\\"left-bottom\\">
  <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content smy-badge__content--fixed smy-badge__content--right-top\\"></span></transition-stub>
  <div style=\\"width: 100px; height: 100px;\\"></div>
</div>"
`;

exports[`test badge style 1`] = `
"<div class=\\"smy-badge smy--box\\">
  <transition-stub name=\\"smy-badge-fade\\"><span class=\\"smy-badge__content smy-badge__content\\" style=\\"top: 100px; right: 1000px; z-index: 500; background: rgb(98, 0, 234);\\"></span></transition-stub>
</div>"
`;
